### PR TITLE
fix: prevent hyphenated project names

### LINKED ISSUE #3687: add auth - SNS IAM role duplicated when project name contains a hyphen "-"
**Describe the bug**
Recent code change assumes project name does not contain a hyphen.
#2709
Commit: fe0e979812bd3142a56a62bf15444cbb3dbbb6bb
File: auth-template.yml.ejs
Code:      RoleName: !If [ShouldNotCreateEnvResources, '<%=`${props.resourceNameTruncated}_sns-role`%>', !Join ['',[ 'sns', !Select [3, !Split ['-', !Ref 'AWS::StackName']], '-', !Ref env]]]

The Select assumes the Split has three parts. With the hyphen involved in the project name the random id looked for selects the environment, which creates a duplicate SNS role name. Duplicated from another project with the same environment.

The use case is create-react-app which doesn't allow capital letters in the project name, so a hyphen was used. The default project name for amplify init then had a hyphen. 

**Amplify CLI Version**
4.16.1

**To Reproduce**
create two directories:
project-1
project-2
In each "amplify init", take the default and give them the same environment.
In each "amplify add auth", take the defaults, amplify push -y

**Expected behavior**
The second push would succeed without issue.

**Screenshots**

**Desktop (please complete the following information):**
OS: Windows
Node Version:12.14.0

**Additional context**

### PR DESCRIPTION
**Pull request by nikhname**
fix #3687

*Issue #, if available:*
#3687

*Description of changes:*
Prevent projects from being created with a hyphenated name

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/config-steps/c0-analyzeProject.js [hunk 1]
```diff
@@ -49,15 +49,15 @@ async function configureProjectName(context) {
 }
 
 function isProjectNameValid(projectName) {
-  return projectName && projectName.length >= 3 && projectName.length <= 20 && /[a-zA-Z0-9]/g.test(projectName);
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/config-steps/c0-analyzeProject.js:52]: if we're going to use this reg ex in multiple places we should declare it as a const somewhere an import it when needed

Comment (edwardfoyle) [reply packages/amplify-cli/src/config-steps/c0-analyzeProject.js:52]: Lol, didn't read the other comments first, looks like this has already been pointed out

**Comment by nikhname** [comment]
> Can we consolidate these to a single function instead of having multiple copies. Also could you verify these are the only places where we have name normalization

Consolidated to a single file and exporting the regex consts for other functions to use. Verified this is the only place with normalization that doesn't have the correct regex

**Comment by nikhname** [comment]
> One small comment.
> 
> Also wouldn't be a bad idea to add some unit tests around the regex as sanity checks

Sure it would definitely be good to have unit tests here with valid and invalid project names

**Comment by nikhname** [comment]
> > One small comment.
> > Also wouldn't be a bad idea to add some unit tests around the regex as sanity checks.
> 
> Sure it would definitely be good to have unit tests here with valid and invalid project names. I'll update here when those are added

Added unit tests that will cover the regex and validation functions

**Comment by mikelhamer** [comment]
The AWS Console still allows projects to be created with "-" in the name. Shouldn't this be consistent?

**Comment by mikelhamer** [comment]
Also can someone please explain to me why this is a thing? Every other AWS resource I can think of allows hyphens, but for an Amplify project ihavetohaveanuglyprojectnamethatishardtoread

**Pull request merged**